core/src/main/java/org/infinispan/interceptors/locking/ClusteringDependentLogic.java

```diff
@@ -508,47 +584,79 @@ protected void commitSingleEntry(CacheEntry entry, Metadata metadata, FlagAffect
                         builder = entry.getMetadata().builder();
                      }
                      metadata = builder
-                           .lifespan(configuration.clustering().l1().lifespan())
-                           .build();
+                        .lifespan(configuration.clustering().l1().lifespan())
+                        .build();
                      metadata = new L1Metadata(metadata);
                   }
-               } else {
-                  doCommit = false;
                }
+               isL1Write = true;
+               doCommit = Commit.COMMIT_NON_LOCAL;
             }
 
-            boolean created = false;
-            boolean removed = false;
-            boolean expired = false;
-            if (!isForeignOwned) {
-               created = entry.isCreated();
-               removed = entry.isRemoved();
-               if (removed && entry instanceof MVCCEntry) {
-                  expired = ((MVCCEntry) entry).isExpired();
+            if (doCommit.isCommit()) {
+               boolean created = false;
+               boolean removed = false;
+               boolean expired = false;
+               if (doCommit.isLocal()) {
+                  created = entry.isCreated();
+                  removed = entry.isRemoved();
+                  if (removed && entry instanceof MVCCEntry) {
+                     expired = ((MVCCEntry) entry).isExpired();
+                  }
                }
-            }
 
-            if (doCommit) {
+               // TODO use value from the entry
                InternalCacheEntry previousEntry = dataContainer.peek(entry.getKey());
                Object previousValue = null;
                Metadata previousMetadata = null;
                if (previousEntry != null) {
                   previousValue = previousEntry.getValue();
                   previousMetadata = previousEntry.getMetadata();
                }
-               commitManager.commit(entry, metadata, trackFlag, l1Invalidation);
-               if (!isForeignOwned) {
-                  notifyCommitEntry(created, removed, expired, entry, ctx, command, previousValue, previousMetadata);
+               if (isL1Write && previousEntry != null && !previousEntry.isL1Entry()) {
+                  // don't overwrite non-L1 entry with L1 (e.g. when originator == backup
+                  // and therefore we have two contexts on one node)
```

Comment:
> CACHE_MODE_LOCAL javadoc explicitly says that write goes to L1. 

Hmm, I never realized that `CACHE_MODE_LOCAL` writes are supposed to go into L1... I think maybe the idea was that `putForExternalRead()` should never go remotely, but it's weird because the node isn't going to be registered as an L1 requestor for that key, so `L1ManagerImpl.flushCache()` isn't going to send any invalidation.

> Pch, I see that runL1UpdateIfPossible writes the L1 entry without DC.compute. I wonder if anyone has ran resilience tests with L1 recently...

Synchronizers have their own per-key locking, I think `DC.compute` isn't needed as long as you don't have `CACHE_MODE_LOCAL` writes to L1.

> But the issue here is different, and I think it was related to the code where I've assumed that puts retrieve the value ahead. There an originator == backup would fetch the remote value into context, execute write on primary and get the value updated, and then it attempted to commit the value it has in context as L1 entry. But that shouldn't be necessary now (I hope) and is prone to the absence of locking anyway.

Yeah, I think now the only way to write to L1 with proper locking is through an `L1Synchronizer`.